drivers/motor_driver/motor_driver.c

```diff
@@ -0,0 +1,209 @@
+            if (gpio_init(motor_driver_conf->motors[i].gpio_enable,
+                GPIO_OUT)) {
+                errno = EIO;
+                LOG_ERROR("gpio_enable init failed\n");
+                goto motor_init_err;
+            }
+            motor_enable(motor_driver, i);
+        }
+    }
+
+    return 0;
+
+motor_init_err:
+    return -errno;
+}
+
+int motor_set(const motor_driver_t motor_driver, uint8_t motor_id, \
+    motor_direction_t direction, uint16_t pwm_duty_cycle)
+{
+    assert(motor_driver < MOTOR_DRIVER_NUMOF);
+
+    const motor_driver_config_t *motor_driver_conf =
+        &motor_driver_config[motor_driver];
+
+    assert(motor_id < motor_driver_conf->nb_motors);
+
+    const motor_t *dev = &motor_driver_conf->motors[motor_id];
+
+    direction = direction ^ dev->gpio_dir_reverse;
+
+    /* Two direction GPIO, handling brake */
+    if (motor_driver_conf->mode == MOTOR_DRIVER_2_DIRS)
+    {
+        if ((dev->gpio_dir0 == GPIO_UNDEF) || \
+            (dev->gpio_dir1_or_brake == GPIO_UNDEF))
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                gpio_write(dev->gpio_dir1_or_brake, direction ^ 0x1);
+                break;
+            case MOTOR_BRAKE:
+            default:
+                /* Brake */
+                gpio_write(dev->gpio_dir0,
+                    motor_driver_conf->mode_brake);
+                gpio_write(dev->gpio_dir1_or_brake,
+                    motor_driver_conf->mode_brake);
+                pwm_duty_cycle = 0;
+                break;
+        }
+    }
+    /* Single direction GPIO */
+    else if (motor_driver_conf->mode == MOTOR_DRIVER_1_DIR)
+    {
+        if (dev->gpio_dir0 == GPIO_UNDEF)
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                break;
+
+            case MOTOR_BRAKE:
+            default:
+                pwm_duty_cycle = 0;
+                break;
+        }
+    }
+    /* Single direction GPIO and brake GPIO */
+    else if (motor_driver_conf->mode == MOTOR_DRIVER_1_DIR_BRAKE)
+    {
+        if ((dev->gpio_dir0 == GPIO_UNDEF) || \
+            (dev->gpio_dir1_or_brake == GPIO_UNDEF))
+        {
+            errno = ENODEV;
+            goto motor_set_err;
+        }
+        switch (direction)
+        {
+            case MOTOR_CW:
+            case MOTOR_CCW:
+                /* Direction */
+                gpio_write(dev->gpio_dir0, direction);
+                /* No brake */
+                gpio_write(dev->gpio_dir1_or_brake, dev->gpio_brake_invert);
```

Comment:
Because our GPIO implementation cannot set more than one pin at the same time, this can glitch. Even worse, the CPU could be interrupted in the between the two calls to gpio_write.

Reply:
Good point I should protect that calls.